src/steamship/experimental/agent_service/agent_service.py

```diff
@@ -0,0 +1,382 @@
+    # (2) agent and tool interactions
+    #
+    # It is meant to be persisted (via a KeyValueStore or otherwise) to allow for
+    # restarts, interruptions, etc.
+
+    # NB: I've forgone the go-style contexts as structs with key-values here for now
+    # i still believe there is value there, but i didn't want that to interfere with
+    # readability too much when just trying to string together the concepts.
+    id: str
+    metadata: Metadata = {}
+
+    # User<->Package chat history (NOT Agent<-->Tool history)
+    # NB: This is distinct from any sort of history related to agent execution
+    # Primarily used to ferry prompts into agent executions AND
+    # to allow for some sort of referential lookup of context as input to the agent
+    chat_history: ChatFile
+
+    # instead of saving the execution history via File as text, this chooses to represent
+    # state as objects. this may not be what we want. This could also be:
+    # agent_history: ChatFile (with tags on blocks to represent state of Tool execution)
+    # i need to think about that a bit more.
+    #
+    # for each of these bits, we probably need to save only IDs, and rehydrate properly;
+    # likely requires an override of `dict()` and `parse_obj()`
+    completed_steps: AgentSteps = []
+
+    # This supports parallel tool execution. But I think that should
+    # be a future iteration, as it complicates things quite a bit.
+    #
+    # as above this likely could be "discovered" from `agent_history: ChatFile` and tags
+    # dynamically, and be more "Steamship-native". Perhaps in a second pass.
+    in_progress: List[
+        Tuple[ToolBinding, Task]
+    ] = []  # todo: should this be a map from task_id -> ToolBinding?
+
+    # this, I think(?), is not serializable, and must be set in some sort of context init bit
+    # of whatever is doing the work.
+    # in the future, this could be a set of callbacks, more broken out (onError, onComplete, ...)
+    emit_funcs: List[EmitFunc] = []
+
+
+class SyncTool(BaseTool, ABC):
+    @abstractmethod
+    def run(self, inputs: List[Block], context: AgentContext) -> List[Block]:
+        pass
+
+
+class AsyncTool(BaseTool, ABC):
+
+    # Here, run() -> Task[List[Block]] might be too limiting
+    # This could be adapted to:
+    # - run() -> Task[Any]
+    # - blockify(Any) -> List[Block]
+    #
+    # run_agent() would then invoke `tool.blockify(tool_task.output)` as an intermediate step
+    # in its control loop once tool_task was complete (saving state to context).
+
+    @abstractmethod
+    def run(self, inputs: List[Block], context: AgentContext) -> Task[List[Block]]:
+        pass
+
+
+class SerpTool(SyncTool, ABC):
+    pass
+
+
+class PodcastDownloaderTool(AsyncTool, ABC):
+    pass
+
+
+class SpeechToTextTool(AsyncTool, ABC):
+    pass
+
+
+class SummarizeTool(AsyncTool, ABC):
+    pass
+
+
+class ImageTool(SyncTool, ABC):
+    pass
+
+
+class ScheduleMessageTool(SyncTool, ABC):
+    pass
+
+
+class Action:
```

Comment:
I really love this metaphor, and it's something that my branch completely lacks and wish it had.

Reply:
@douglas-reid am I taking this in a different direction than you're intending it when I start to think of a ChatHistory as having blocks tagged with `Action` in addition to `Message`?

The `Action` block have the JSON-serialized version of this object?

Reply:
@eob i couldn't decide if `Action` persists, or not, but I don't think that's off base. that's something i started to think about after i sorta got things ripped apart: a la if you reuse a ChatHistory concept, *just for localized agent history/scratchpad*, can we dump the history via blocks with tags via something like `Action`?